### PR TITLE
[platform] Stream chunk downloads to disk instead of buffering in RAM

### PR DESCRIPTION
**Pull request by biodranik**
- Fixes an OOM on Android
- Separates segments/chunks download on all platforms
- Stricter chunk validation and corner-case handling
- Tests coverage
- Updated gradle to the latest version to fix errors in the latest Android Studio

### CODE REVIEW

**Comment by vng** [comment]
Did a quick look. Is it reasonable to avoid duplication logic in processSegmentResponse, validateSegmentResponse for the platforms (with inner parseContentRange)?

File: android/sdk/src/main/cpp/app/organicmaps/sdk/util/HttpClient.cpp [hunk 1]
```diff
@@ -276,7 +291,13 @@ extern "C" JNIEXPORT void Java_app_organicmaps_sdk_util_HttpClient_nativeOnCompl
...
   {
     result = ExtractResult(env, ctx->m_paramsGlobalRef);
-    result.m_success = true;
+    // Negative error codes (kInconsistentFileSize, kWriteException) signal an
+    // application-level failure even though the HTTP transfer completed without
+    // throwing. Segment mode is stricter: only 206 means the requested chunk was
+    // accepted and written successfully — any other HTTP code (404, 416, 500, ...)
+    // leaves no bytes on disk, so must not be reported as success. Matches Qt/Apple
+    // m_success semantics.
+    result.m_success = result.m_errorCode >= 0 && (!ctx->m_isSegmentMode || result.m_errorCode == 206);
```
Comment (vng) [inline android/sdk/src/main/cpp/app/organicmaps/sdk/util/HttpClient.cpp:313]: ```ctx->m_isSegmentMode ? result.m_errorCode == 206 : result.m_errorCode == 200``` ?

Comment (biodranik) [reply android/sdk/src/main/cpp/app/organicmaps/sdk/util/HttpClient.cpp:313]: They are not equivalent.

**Comment by biodranik** [comment]
> Is it reasonable to avoid duplication logic in processSegmentResponse, validateSegmentResponse for the platforms (with inner parseContentRange)?

Please check the last commit, if it's easier, we can leave it.

File: android/sdk/src/main/cpp/app/organicmaps/sdk/util/HttpClient.cpp (hunk 1 above)
Comment (vng) [reply android/sdk/src/main/cpp/app/organicmaps/sdk/util/HttpClient.cpp:313]: You check "result.m_errorCode == 200" in C++ code, so I mean probably you forgot here.

**Comment by vng** [comment]
nit: We used to write
```
extern "C"
{
}
```
for the whole module, not every function.

**Comment by vng** [comment]
As usual, can't say for sure whether it is applicable here or not.
-------------------------------------------------------------
**Findings**
- Medium: segment mode bypasses the shared streaming path, so any caller that combines `SetReceivedFileSegment` with `ProgressHandler` or `DataHandler` silently loses those callbacks. In HttpClient.java the method returns into `processSegmentResponse(...)` before `readBodyStreaming(...)`, and the Qt backend follows the same one-off body handling in http_client_qt.cpp. That is a routing regression and makes the new mode behave differently from the rest of `HttpClient`.

- Low/medium: the PR duplicates the same segment validation contract across Java, Qt, and the shared C++ helper instead of centralizing the file-write/range logic. The overlap between HttpClient.java, http_client_qt.cpp, and http_client.cpp is already large enough that any future tweak to `206` / `Content-Range` / error-code handling will need to be changed in multiple places. That is unnecessary overhead and the main copy-paste risk here.

No other blocking correctness issues stood out.

File: android/sdk/src/main/cpp/app/organicmaps/sdk/util/HttpClient.cpp (hunk 1 above)
Comment (biodranik) [reply android/sdk/src/main/cpp/app/organicmaps/sdk/util/HttpClient.cpp:313]: Double-checked, it's ok. Semantic is:
  - non-segment 404 => m_success == true
  - segment 206 => m_success == true
  - segment 404 => m_success == false
  - segment kInconsistentFileSize / kWriteException => m_success == false

**Comment by biodranik** [comment]
@vng not critical (not our case), but just in case:
- Call progress callback as on iOS/Qt
- CHECK to avoid mixing incompatible modes
- nit fixed

**Comment by biodranik** [comment]
@vng should be ready to merge. As always, some additional testing doesn't hurt.

**Review by vng — APPROVED** [review]
